### ISSUE TITLE
[$250] When there is only one message in a chat then on deleting that message it shows "_fake_" in LHN - reported by @Tushu17

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8361]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Click on :green-plus-button:
2. New chat > Search a new user
3. Message the searched user
4. Delete the message and check the LHN

## Expected Result:
It shouldn't show the "_fake_" in LHN

## Actual Result:
LHN shows "_fake_"

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.46-3
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/160738760-46bb5613-684b-421a-82e8-5d723da17a6f.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @Tushu17
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1647727189086589

**Job Post:** https://www.upwork.com/jobs/~0182dfea977510a8eb

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by kevinksullivan** [comment]
@Tushu17 please apply to the posted job for reporting bonus. Thanks!

**Comment by brianmuks** [comment]
**Proposal**
The bug is coming from this line:
https://github.com/Expensify/App/blob/ce6363cd7d08819c6ea94b7ffb6a8009eeab3775/src/libs/actions/ReportActions.js#L105

```
   const lastMessageIndex = _.findLastIndex(reportActions[reportID], action => (
        !ReportUtils.isDeletedAction(action)
    ));
```
 This does not happen when you  just create one message and delete it. It also happens when you delete all the messages.
 If I am not mistaken ```reportActions[reportID]``` always contains an object at index 0 with this data:

<img width="746" alt="Screenshot 2022-04-02 at 11 15 59 PM" src="https://user-images.githubusercontent.com/4520893/161401484-fa4f5374-ffef-4203-b722-27bd4b7e90bb.png">

So when all the messages are  deleted the object at index 0 is the one that will pass this test :
https://github.com/Expensify/App/blob/ce6363cd7d08819c6ea94b7ffb6a8009eeab3775/src/libs/actions/ReportActions.js#L106

Here much depends with what the expected results are. Here is my solution. 
**Solution**
When ```lastMessageIndex``` returns ```0```  we should get the index of the last message from ```reportActions[reportID]``` 
```
    let lastMessageIndex = _.findLastIndex(reportActions[reportID], action => (
        !ReportUtils.isDeletedAction(action)
    ));

    lastMessageIndex = lastMessageIndex ? lastMessageIndex : reportActions[reportID].length
```
https://github.com/Expensify/App/blob/ce6363cd7d08819c6ea94b7ffb6a8009eeab3775/src/libs/actions/ReportActions.js#L105






https://user-images.githubusercontent.com/4520893/161401735-623f6350-aa9d-4b70-818f-c1fee9f76f72.mov

**Comment by Santhosh-Sellavel** [comment]
I will review the proposal(s) by tomorrow.

**Comment by Santhosh-Sellavel** [comment]
@thienlnam What's expected here?

**Comment by thienlnam** [comment]
I think it showing either blank or some sort of indication the message was deleted works here

**Comment by Santhosh-Sellavel** [comment]
Thanks, @thienlnam.
I think it should be blank because we don't show any deleted indication if delete the last message. Instead, the message that was before deleted is shown in the preview.

@brianmuks Can you update your proposal to match the expected OP let us know thanks!

**Comment by brianmuks** [comment]
> Thanks, @thienlnam. I think it should be blank because we don't show any deleted indication if delete the last message. Instead, the message that was before deleted is shown in the preview.
> 
> @brianmuks Can you update your proposal to match the expected OP let us know thanks!

Thanks for the feedback. My proposal at the moment shows blank but what I understand is that you want the last deleted message to still show on the LHN . Is that right ?

**Comment by Santhosh-Sellavel** [comment]
No @brianmuks It should be blank.

**Comment by Santhosh-Sellavel** [comment]
Will this https://github.com/Expensify/App/issues/8361#issuecomment-1086727568 show a blank on LHN? @brianmuks

**Comment by brianmuks** [comment]
> Will this [#8361 (comment)](https://github.com/Expensify/App/issues/8361#issuecomment-1086727568) show a blank on LHN? @brianmuks

Yes it will show blank.

**Comment by brianmuks** [comment]
> No @brianmuks It should be blank.

Then this proposal aligns well with what you are looking for https://github.com/Expensify/App/issues/8361#issuecomment-1086727568

**Comment by brianmuks** [comment]
> @thienlnam What's expected OP here?

I am catching up here. What does OP stand for ? I want to sound like an engineer too 😄

**Comment by Santhosh-Sellavel** [comment]
@brianmuks 

Your proposal is not clear, I am not able to follow your **Solution** can brief it more clearly, and instead of the diff(which is confusing) can you send a snippet of the modified block.

**Comment by Santhosh-Sellavel** [comment]
OP means Output

**Comment by thienlnam** [comment]
Let me check with the team to see verify what the expected output should be.  Will post back here when I hear back

**Comment by Santhosh-Sellavel** [comment]
Okay @thienlnam

**Comment by brianmuks** [comment]
> I'm not following here, `lastMessageIndex = lastMessageIndex ? lastMessageIndex : reportActions[reportID].length`
> 
> @brianmuks Can you briefly explain how this addresses the issue?

My comment here explains the technical aspect: https://github.com/Expensify/App/issues/8361#issuecomment-1086727568

```
lastMessageIndex = lastMessageIndex ? lastMessageIndex : reportActions[reportID].length
```
The snippet above ensures that the index  of  object where the ```_fake_``` is coming from is not used as the last message in the LHN.  From my understanding ```reportActions[reportID][0]``` contains auto-generated data.

**Comment by Santhosh-Sellavel** [comment]
`reportActions[reportID].length` Does make any sense to me.

**Comment by brianmuks** [comment]
> `reportActions[reportID].length` Does make any sense to me.

I see your point. I was trying to get the index  of the last deleted message . The correction then should be ```reportActions[reportID].length -1 ```.

**Comment by brianmuks** [comment]
> > `reportActions[reportID].length` Does make any sense to me.
> 
> I see your point. I was trying to get the index of the last deleted message . The correction then should be `reportActions[reportID].length -1 ` but this will break if you send only one message then you delete it .
> 
> If 3 messages were deleted `reportActions[reportID].length` will return `4`. Off-course there is no message at index `4`. The snippet below handles such a scenario especially if you start a new chat, send one message then you delete it. I don't think anything will break.
> 
> ```
>     return ReportUtils.formatReportLastMessageText(
>         lodashGet(reportActions, [reportID, lastMessageIndex, 'message', 0, 'text'], ''),
>     );
> ```
> 
> Alternatively we can just do this :
> 
> ```
>     return lastMessageIndex ? ReportUtils.formatReportLastMessageText(
>         lodashGet(reportActions, [reportID, lastMessageIndex, 'message', 0, 'text'], ''),
>     ) : '';
> ```
> 
> Return an empty string when lastMessageIndex is `0`.

@Santhosh-Sellavel any comment on this ?

**Comment by thienlnam** [comment]
Pending review from C+.  I don't think we should be changing what the lastMessageIndex is, it makes more sense to return an empty string if the lastMessageIndex is 0

**Comment by brianmuks** [comment]
@thienlnam how is this :
```
 return lastMessageIndex ? ReportUtils.formatReportLastMessageText(
         lodashGet(reportActions, [reportID, lastMessageIndex, 'message', 0, 'text'], ''),
     ) : '';
```

**Comment by Santhosh-Sellavel** [comment]
>  I don't think we should be changing what the lastMessageIndex is

I agree with you here @thienlnam.

But I would dig little this a bit before moving forward, I'll update on Monday thanks!

**Comment by Santhosh-Sellavel** [comment]
@thienlnam 

@brianmuks [Proposal](https://github.com/Expensify/App/issues/8361#issuecomment-1106881771) Looks Good!
🎀 👀 🎀 C+ reviewed

**Comment by kevinksullivan** [comment]
Hired @brianmuks , @Santhosh-Sellavel , and @Tushu17 for the job in upwork. Dream team!

**Comment by Santhosh-Sellavel** [comment]
@brianmuks Any update here ?

**Comment by Santhosh-Sellavel** [comment]
@kevinksullivan @thienlnam 
As per [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) guidelines,

> 11. Daily updates on weekdays are highly recommended. If you know you won’t be able to provide updates for > 1 week, please comment on the PR or issue how long you plan to be out so that we may plan accordingly. We understand everyone needs a little vacation here and there. Any issue that doesn't receive an update for 1 full week may be considered abandoned and the original contract terminated.

There is no update over one week on this one, can we terminate @brianmuks contract add the Help Wanted Label here?

**Comment by brianmuks** [comment]
> @kevinksullivan @thienlnam As per [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) guidelines,
> 
> > 11. Daily updates on weekdays are highly recommended. If you know you won’t be able to provide updates for > 1 week, please comment on the PR or issue how long you plan to be out so that we may plan accordingly. We understand everyone needs a little vacation here and there. Any issue that doesn't receive an update for 1 full week may be considered abandoned and the original contract terminated.
> 
> There is no update over one week on this one, can we terminate @brianmuks contract add the Help Wanted Label here?

Sorry for the delay. PR will be ready by tomorrow 🙏

**Comment by kevinksullivan** [comment]
Hi @brianmuks , do you have an update/PR for us today?

**Comment by brianmuks** [comment]
> Hi @brianmuks , do you have an update/PR for us today?

@kevinksullivan PR is ready **https://github.com/Expensify/App/pull/8967**

**Comment by thienlnam** [comment]
We're going to be revisiting a solution for group chats since currently with the fix it shows the name of the user that last sent a message and then no message which seems like a bug.

**Comment by brianmuks** [comment]
> We're going to be revisiting a solution for group chats since currently with the fix it shows the name of the user that last sent a message and then no message which seems like a bug.

Further to my proposal. To fix the mentioned bug ☝🏾 we can check if the last message is an empty string. If so then we can just set `lastMessageText` to an empty string which in turn will force LHN to show the `subtitle` instead of the name of the user who sent the last deleted message as it is the case currently. 
https://github.com/Expensify/App/blob/8e3c67ba25200f5558ce6be6fd3aa211c43d60b6/src/libs/OptionsListUtils.js#L237
```  
 let lastMessageText = report && !isArchivedRoom && hasMultipleParticipants && lastActorDetails && lodashGet(report, 'lastMessageText', null)
```

https://user-images.githubusercontent.com/4520893/170326162-debe0a3c-2732-48ce-9a45-96c8c91f94eb.mov

**Comment by Santhosh-Sellavel** [comment]
@brianmuks How it works out for group chat?

**Comment by brianmuks** [comment]
> @brianmuks How it works out for group chat?
> 
> Can you share screen recording covering both before adding the message case & after deleting the message.
@Santhosh-Sellavel here you go : 

https://user-images.githubusercontent.com/4520893/170452812-73504dd1-20f6-41d2-9f9d-7fbc2fa79898.mov

**Comment by Santhosh-Sellavel** [comment]
@thienlnam 
@brianmuks new changes look good to me.

One suggestion there are too many checks for one boolean, can we break down checks in a meaningful way?

**Comment by brianmuks** [comment]
> && lodashGet(report, 'lastMessageText', null)

Thanks @Santhosh-Sellavel .
How is this  ?
https://github.com/Expensify/App/blob/8e3c67ba25200f5558ce6be6fd3aa211c43d60b6/src/libs/OptionsListUtils.js#L237
```  
    let lastMessageText = report && hasMultipleParticipants && lastActorDetails  ? `${lastActorDetails.displayName}:  : ''; `
    //Since no empty message can be sent we next  check if the message was deleted 
   lastMessageText =  lodashGet(report, 'lastMessageText', null) ? lastMessageText : '';
```

**Comment by thienlnam** [comment]
You're somewhat on the right track, but you're removing the subtitle name completely which is not what we want to do here.  

Also, in the video you linked - the expected functionality is incorrect.  If you're deleting the only message in the group chat, the subtitle should be empty and not show anything

**Comment by brianmuks** [comment]
@thienlnam I am not removing the subtitle you can check the video below or I could be missing something here. 
I had this comment below from @Santhosh-Sellavel  when I revisited the solution. The aim has been to restore all chats to their original state as when one initiated the conversation. 
>Additional Context:
Both One to One & group chats are restored to the same state before no messages after deleting the sent messages.
i.e email is shown refer to image/video above


https://user-images.githubusercontent.com/4520893/171154254-253229dc-0ede-4982-be2f-494c9c803a48.mov

**Comment by Santhosh-Sellavel** [comment]
@brianmuks 

On this line https://github.com/Expensify/expensify-common/pull/451
@thienlnam has a different requirement for group chat alone. Not to show email at all.

**Comment by Santhosh-Sellavel** [comment]
@brianmuks 

On this line https://github.com/Expensify/expensify-common/pull/451
@thienlnam has a different requirement for group chat alone. Not to show email at all.

**Comment by brianmuks** [comment]
Thanks @Santhosh-Sellavel for the clarification 🙏🏾 . Let me work on it.

**Comment by brianmuks** [comment]
@thienlnam I was checking if there is a method to check if chat is a group chat but it seems there is none. 
It was suggested that I create one on the thread below. I would like to hear your thoughts on this . 
https://expensify.slack.com/archives/C01GTK53T8Q/p1654092407682469

**Comment by brianmuks** [comment]
I have created a method to determine if chat is group chat as suggested here :
https://expensify.slack.com/archives/C01GTK53T8Q/p1654093853902489?thread_ts=1654092407.682469&cid=C01GTK53T8Q
```
 * Whether the provided report is a group chat
 * @param {Object} report
 * @param {String} report.chatType
 * @returns {Boolean}
 */
function isGroupChat(report) {
    return report.participants.length > 2;
}
```
The methods is sitting here:
https://github.com/brianmuks/Expensify.cash/blob/2a8b8a490e224d6e0cd02bbf39853dd9c4fdff24/src/libs/reportUtils.js
Can I go ahead with this approach ?

**Comment by thienlnam** [comment]
It might be slightly more complicated and you will need to check if the report participants are still greater than 1 if you exclude these emails https://github.com/Expensify/App/blob/main/src/pages/NewChatPage.js#L63-L63

**Comment by brianmuks** [comment]
> It might be slightly more complicated and you will need to check if the report participants are still greater than 1 if you exclude these emails https://github.com/Expensify/App/blob/main/src/pages/NewChatPage.js#L63-L63

Thanks @thienlnam . I am on it !

**Comment by JosueEchandiaAsto** [comment]
@thienlnam I created a method in this [proposal](https://github.com/Expensify/App/issues/9238#issuecomment-1159137959) that returns true if the session email is the only participant. With that I can fix the missing logic in this problem. Can I elaborate some proposal for this problem?

**Comment by Santhosh-Sellavel** [comment]
@JosueEchandiaAsto We already hired someone to do this job.

**Comment by Santhosh-Sellavel** [comment]
@brianmuks Any update on this?

**Comment by brianmuks** [comment]
@Santhosh-Sellavel I am pushing an update today. Sorry for the delay 😞 . I was having issues with the app it couldn't function properly. I think it had to do with the caching issues.

**Comment by brianmuks** [comment]
> It might be slightly more complicated and you will need to check if the report participants are still greater than 1 if you exclude these emails https://github.com/Expensify/App/blob/main/src/pages/NewChatPage.js#L63-L63

@thienlnam it appears that even when we exclude the suggested `EXPENSIFY_EMAILS`, for #annouce chat room we might still get participants being greater than one as seen in the screen shot.
 I am still trying to figure out the best approach to detect a group chat. 

<img width="1269" alt="Screenshot 2022-06-26 at 5 42 33 PM" src="https://user-images.githubusercontent.com/4520893/175822398-20bb78a9-8ee4-41b4-be52-d59f227a368b.png">

**Comment by brianmuks** [comment]
Why can't we have something like this for group chat:
```
function isGroupChat(report) {
    return lodashGet(report, ['chatType'], '') === CONST.REPORT.CHAT_TYPE.POLICY_GROUP_CHAT;
}
```
This means that the following will have to be modified to include `POLICY_GROUP_CHAT`
https://github.com/Expensify/App/blob/406040d1b208277e44580bb372323e934b112f37/src/CONST.js#L247
```
     CHAT_TYPE: {
            POLICY_ANNOUNCE: 'policyAnnounce',
            POLICY_ADMINS: 'policyAdmins',
            DOMAIN_ALL: 'domainAll',
            POLICY_ROOM: 'policyRoom',
            POLICY_EXPENSE_CHAT: 'policyExpenseChat',
           +POLICY_GROUP_CHAT: 'policyGroupChat',
        },
```

**Comment by Santhosh-Sellavel** [comment]
@thienlnam @brianmuks  

> It might be slightly more complicated and you will need to check if the report participants are still greater than 1 if you exclude these emails https://github.com/Expensify/App/blob/main/src/pages/NewChatPage.js#L63-L63

We won't able to create a chat with the excluded emails.

Policy rooms or default rooms have value in `chatType`. So if `chatType` is empty and participants are greater than > 1 in reports. This check should be fine right?

**Comment by brianmuks** [comment]
> @brianmuks What's the goal of this method / how will it solve the solution?

The goal is to use the method to determine if chat is group chat. If true based on this https://github.com/Expensify/App/issues/8361#issuecomment-1142078639 we should be able to 
show nothing when a group chat is initiated and when all messages are deleted.

**Comment by brianmuks** [comment]
> Okay, in this case let's just check if there are multiple participants including expensify emails. If there are more than 2 people in a group chat, it will append the name of the user

If I am getting this right:
1. We don't need a method to determine if chat is group chat
 > it will append the name of the user
2. For group chats : Instead of showing nothing when all messages are deleted we should show the name of the user who sent the last message

**Comment by thienlnam** [comment]
> We don't need a method to determine if chat is group chat

**Comment by brianmuks** [comment]
@thienlnam thanks for clarifying this. I am on it. 
>So right now, our issue is that for chats with multiple participants - it will show the name of the person sending the message right?
Yes

**Comment by Santhosh-Sellavel** [comment]
Any update @brianmuks?

**Comment by brianmuks** [comment]
This issue https://github.com/Expensify/App/issues/8361#issuecomment-1164548104 keeps persisting. I think it might be a bug. I have tried to clear cookies to no avail. Let me check on other platforms.

**Comment by thienlnam** [comment]
> This issue https://github.com/Expensify/App/issues/8361#issuecomment-1164548104 keeps persisting. I think it might be a bug. I have tried to clear cookies to no avail. Let me check on other platforms.

We must have fixed the _fake_ in another change.  However, if you have two messages and delete the most recent one, it correctly shows the message sent before correct?

**Comment by brianmuks** [comment]
> We must have fixed the _fake_ in another change. 

Yes the _fake_ issue was fixed but not merged yet. 

> However, if you have two messages and delete the most recent one, it correctly shows the message sent before correct?
No, the deleted messages are not disappearing on the LHN only from the chat view

**Comment by thienlnam** [comment]
> Yes the fake issue was fixed but not merged yet.

What do you mean by not merged yet - you're experiencing it on every branch so it must be on main right? Do you mean not deployed?

> No, the deleted messages are not disappearing on the LHN only from the chat view

I just tested deleting a message on prod (Version 1.1.82-5 (1.1.82-5).  It will show the last message

It looks like this issue is resolved - the outcome we've had is also acceptable (showing the email if the only message is deleted)

**Comment by Santhosh-Sellavel** [comment]
Yes, `_fake_` issue is no longer reproducible now in production. Even though @brianmuks PR #8967 is not merged yet.

**Comment by thienlnam** [comment]
@kevinksullivan Let's pay out this issue to pay them for their time but this issue has been fixed by another PR

**Comment by kevinksullivan** [comment]
Paid @Santhosh-Sellavel and @Tushu17 .  @brianmuks please accept hire.

**Comment by brianmuks** [comment]
> Paid @Santhosh-Sellavel and @Tushu17 . @brianmuks please accept hire.
Done. Thanks

**Comment by Santhosh-Sellavel** [comment]
@thienlnam for group name case shall we create a new issue or shall we update the description and reapply the Help Wanted label?

**Comment by thienlnam** [comment]
Current functionality is fine - we can leave it as that for now

**Comment by brianmuks** [comment]
> @brianmuks please accept hire.
@kevinksullivan I am waiting on you 😄 . Thanks.
